expression/builtin_compare.go

```diff
@@ -1070,68 +1070,80 @@ func isTemporalColumn(expr Expression) bool {
 }
 
 // tryToConvertConstantInt tries to convert a constant with other type to a int constant.
-func tryToConvertConstantInt(ctx sessionctx.Context, isUnsigned bool, con *Constant) (_ *Constant, isAlwaysFalse bool) {
+// The second return-val only used when isAlways is True
+func tryToConvertConstantInt(ctx sessionctx.Context, targetFieldType *types.FieldType, con *Constant) (_ *Constant, _ *Constant, isAlways bool) {
 	if con.GetType().EvalType() == types.ETInt {
-		return con, false
+		return con, nil, false
 	}
 	dt, err := con.Eval(chunk.Row{})
 	if err != nil {
-		return con, false
+		return con, nil, false
 	}
 	sc := ctx.GetSessionVars().StmtCtx
-	fieldType := types.NewFieldType(mysql.TypeLonglong)
-	if isUnsigned {
-		fieldType.Flag |= mysql.UnsignedFlag
-	}
+	fieldType := targetFieldType
+
 	dt, err = dt.ConvertTo(sc, fieldType)
 	if err != nil {
-		return con, terror.ErrorEqual(err, types.ErrOverflow)
+		if terror.ErrorEqual(err, types.ErrOverflow) {
+			return con, &Constant{
+				Value:        dt,
+				RetType:      fieldType,
+				DeferredExpr: con.DeferredExpr,
+			}, true
+		}
+		return con, nil, false
 	}
 	return &Constant{
 		Value:        dt,
 		RetType:      fieldType,
 		DeferredExpr: con.DeferredExpr,
-	}, false
+	}, nil, false
 }
 
 // RefineComparedConstant changes an non-integer constant argument to its ceiling or floor result by the given op.
-// isAlwaysFalse indicates whether the int column "con" is false.
-func RefineComparedConstant(ctx sessionctx.Context, isUnsigned bool, con *Constant, op opcode.Op) (_ *Constant, isAlwaysFalse bool) {
+// isAlways indicates whether the int column "con" is true/false.
+// The second return-val only used when isAlways is True
+func RefineComparedConstant(ctx sessionctx.Context, targetFieldType *types.FieldType, con *Constant, op opcode.Op) (_ *Constant, _ *Constant, isAlways bool) {
```

Comment:
Maybe we should pass types.FieldType instead of *types.FieldType here to avoid it be modified wrongly.